storage/artifactory.go

```diff
@@ -262,28 +267,46 @@ func (s *Artifactory) AddExtension(ctx context.Context, manifest *VSIXManifest,
 		return "", err
 	}
 
+	for _, file := range extra {
+		// TODO: I think this is correct?
+		_, err := s.upload(ctx, path.Join(dir, file.RelativePath), bytes.NewReader(file.Content))
+		if err != nil {
+			return "", err
+		}
+	}
+
 	return s.uri + dir, nil
 }
 
-func (s *Artifactory) FileServer() http.Handler {
-	// TODO: Since we only extract a subset of files perhaps if the file does not
-	// exist we should download the vsix and extract the requested file as a
-	// fallback.  Obviously this seems like quite a bit of overhead so we would
-	// then emit a warning so we can notice that VS Code has added new asset types
-	// that we should be extracting to avoid that overhead.  Other solutions could
-	// be implemented though like extracting the VSIX to disk locally and only
-	// going to Artifactory for the VSIX when it is missing on disk (basically
-	// using the disk as a cache).
-	return http.HandlerFunc(func(rw http.ResponseWriter, r *http.Request) {
-		reader, code, err := s.read(r.Context(), r.URL.Path)
-		if err != nil {
-			http.Error(rw, err.Error(), code)
-			return
+// Open returns a file from Artifactory.
+// TODO: Since we only extract a subset of files perhaps if the file does not
+// exist we should download the vsix and extract the requested file as a
+// fallback.  Obviously this seems like quite a bit of overhead so we would
+// then emit a warning so we can notice that VS Code has added new asset types
+// that we should be extracting to avoid that overhead.  Other solutions could
+// be implemented though like extracting the VSIX to disk locally and only
+// going to Artifactory for the VSIX when it is missing on disk (basically
+// using the disk as a cache).
+func (s *Artifactory) Open(ctx context.Context, fp string) (fs.File, error) {
+	resp, code, err := s.request(ctx, http.MethodGet, path.Join(s.repo, fp), nil)
+	if err != nil {
+		switch code {
+		case http.StatusNotFound:
+			return nil, fs.ErrNotExist
+		case http.StatusForbidden:
+			return nil, fs.ErrPermission
+		default:
+			return nil, err
 		}
-		defer reader.Close()
-		rw.WriteHeader(http.StatusOK)
-		_, _ = io.Copy(rw, reader)
-	})
+	}
+
+	f := mem.NewFileHandle(mem.CreateFile(fp))
```

Comment:
I was really wondering about this too, but ended up not spending the time to try to fix it.

> Or...is it possible to implement a File that Reads from the HTTP response? Not sure how Stat would be implemented though. In any case, probably not a big deal. The biggest file is probably the .vsix, and surely there are not multi-gigabyte .vsix files I would hope.

This would be ideal, you can implement `Stat` rather easily, it's implementing the `io.Seeker` that was problematic when I spent a few minutes on it.

I think we should revisit this and try to fix it. I'll throw on a comment. If we re-implement `http.FileServer`, we might be able to make our own easier to implement interface too. Ignore things like `Stat`

Reply:
Awesome, that sounds good to me.